pkg/controller/wip/controller.go

```diff
@@ -119,8 +125,151 @@ func (c *controller) brokerDelete(obj interface{}) {
+	brokerClient := c.brokerClientCreateFunc(broker.Name, broker.Spec.URL, username, password)
+	brokerCatalog, err := brokerClient.GetCatalog()
+	if err != nil {
+		glog.Errorf("Error getting broker catalog for broker %v: %v", broker.Name, err)
+		c.updateBrokerReadyCondition(broker, v1alpha1.ConditionFalse, ErrorFetchingCatalogReason, ErrorFetchingCatalogMessage)
+		return
+	} else {
+		glog.V(5).Infof("Successfully fetched %v catalog entries for Broker %v", len(brokerCatalog.Services), broker.Name)
+	}
+
+	glog.V(4).Infof("Converting catalog response for Broker %v into service-catalog API", broker.Name)
+	catalog, err := convertCatalog(brokerCatalog)
+	if err != nil {
+		glog.Errorf("Error converting catalog payload for broker %v to service-catalog API: %v", broker.Name, err)
+		err = c.updateBrokerReadyCondition(broker, v1alpha1.ConditionFalse, ErrorSyncingCatalogReason, ErrorSyncingCatalogMessage)
+		if err != nil {
+			glog.Errorf("Error updating ready condition for Broker %v: %v", broker.Name, err)
+		} else {
+			glog.V(5).Infof("Updated ready condition for Broker %v to False", broker.Name)
+		}
+
+		return
+	} else {
+		glog.V(5).Infof("Successfully converted catalog payload from Broker %v to service-catalog API", broker.Name)
+	}
+
+	for _, serviceClass := range catalog {
+		glog.V(4).Infof("Reconciling serviceClass %v (broker %v)", serviceClass.Name, broker.Name)
+		if err := c.reconcileServiceClassFromBrokerCatalog(broker, serviceClass); err != nil {
+			glog.Errorf("Error reconciling serviceClass %v (broker %v): %v", serviceClass.Name, broker.Name, err)
+			c.updateBrokerReadyCondition(broker, v1alpha1.ConditionFalse, ErrorSyncingCatalogReason, ErrorSyncingCatalogMessage)
+			return
+		} else {
+			glog.V(5).Infof("Reconciled serviceClass %v (broker %v)", serviceClass.Name, broker.Name)
+		}
+	}
+
+	err = c.updateBrokerReadyCondition(broker, v1alpha1.ConditionTrue, "FetchedCatalog", "Successfully fetched catalog from broker")
+	if err != nil {
+		glog.Errorf("Error updating ready condition for Broker %v: %v", broker.Name, err)
+	} else {
+		glog.V(5).Infof("Updated ready condition for Broker %v to True", broker.Name)
+	}
+}
+
+// reconcileServiceClassFromBrokerCatalog reconciles a ServiceClass after the
+// Broker's catalog has been re-listed.
+func (c *controller) reconcileServiceClassFromBrokerCatalog(broker *v1alpha1.Broker, serviceClass *v1alpha1.ServiceClass) error {
+	serviceClass.BrokerName = broker.Name
+
+	existingServiceClass, err := c.serviceClassLister.Get(serviceClass.Name)
+	if err != nil {
+		// An error returned from a lister Get call means that the object does
+		// not exist.  Create a new ServiceClass.
+		if _, err := c.serviceCatalogClient.ServiceClasses().Create(serviceClass); err != nil {
+			glog.Errorf("Error creating serviceClass %v from Broker %v: %v", serviceClass.Name, broker.Name, err)
+			return err
+		}
+	}
+
+	clone, err := api.Scheme.DeepCopy(existingServiceClass)
```

Comment:
You're gonna have problems here. On line 197, you're _likely_ to encounter the condition where the `ServiceClasss` doesn't exist, so `existingServiceClass` will be `nil` and you'll fall into the error condition wherein you go ahead an create it the missing `ServiceClass`, but when you reach line 207, `existingServiceClass` is _still_ `nil`.